### ISSUE TITLE
NettyTcpClient won't use new connect address for reconnecting

### ISSUE THREAD

**Issue by shanexu** [reactor/reactor#379]
When reconnect provides a new connect address, NettyTcpClient won't use it for reconnecting.
I guess tow parts of the file "reactor/reactor-net/src/main/java/reactor/net/netty/tcp/NettyTcpClient.java" cause the problem.
```java
private void attemptReconnect(Tuple2<InetSocketAddress, Long> tup) {
			connectAddress = tup.getT1(); /* actually, it is ReconnectingChannelListener.connectAddress = tup.getT1(); */
			bootstrap.remoteAddress(connectAddress);
			long delay = tup.getT2();

			if (log.isInfoEnabled()) {
				log.info("Failed to connect to {}. Attempting reconnect in {}ms.", connectAddress, delay);
			}

			getEnvironment().getRootTimer()
			                .submit(
					                new Consumer<Long>() {
						                @Override
						                public void accept(Long now) {
							                openChannel(ReconnectingChannelListener.this);
						                }
					                },
					                delay,
					                TimeUnit.MILLISECONDS
			                )
			                .cancelAfterUse();
		}
```
```java
		this.connectionSupplier = new Supplier<ChannelFuture>() {
			@Override
			public ChannelFuture get() {
				if (!closing) {
					return bootstrap.connect(getConnectAddress()); /* it always return supers(TcpClient)'s connectAddress, which is assigned during constructing.  */
				} else {
					return null;
				}
			}
		};
```

**Comment by romanek-adam** [comment]
Hi,

my team has probably been hit by this issue, but I want to be sure. Our scenarios is as follows.

We use Spring Messaging STOMP client which in turn [uses NettyTcpClient](https://github.com/spring-projects/spring-framework/blob/master/spring-messaging/src/main/java/org/springframework/messaging/tcp/reactor/Reactor2TcpClient.java#L50). In our scenario we initially provide it a DNS name, e.g. foobar.com which at that time is resolved to lets say 192.168.1.100. Then after certain amount of time we stop the server running on 192.168.1.100, start a new one on another IP address and re-point the DNS name foobar.com to the new address. Unfortunately the NettyTcpClient still tries to connect to the old address which is accompanied by the following log entry:

```2015-11-16 12:11:15.855  INFO 1 --- [eactor-tcp-io-2] r.io.net.impl.netty.tcp.NettyTcpClient   : Failed to connect to foobar.com/192.168.1.100:61613. Attempting reconnect in 5000ms.```

The message is repeated over and over again and the DNS name is never resolved again (I left a test for 24h, no change).

In general this issue prevents us from implementing a true HA service as DNS-based failover does not work.

Is it a bug in NettyTcpClient or am I missing something?

**Comment by rstoyanchev** [comment]
@smaldini it looks we'll need a fix for this.

First a step back for [SPR-13702](https://jira.spring.io/browse/SPR-13702) one could argue that if `Spec` is called with [connect(host, port)](https://github.com/reactor/reactor/blob/2.0.x/reactor-net/src/main/java/reactor/io/net/Spec.java#L250) is called vs the variant taking an `InetSocketAddress` then the resulting `ReactorClient` knows enough to supply a freshly re-resolved `InetSocketAddress` on reconnect attempts.

I tried to work around that by holding on the host and port and ignoring the reconnect address passed into the `Reconnect` strategy. However even then from what I can see `NettyTcpClient` cannot reconnect to a different address which is the issue reported on this ticket. The call to [boostrap.remoteAddress(connectAddres)](https://github.com/reactor/reactor/blob/2.0.x/reactor-net/src/main/java/reactor/io/net/impl/netty/tcp/NettyTcpClient.java#L334) with the new address only affects calls to the no-arg `bootstrap.connect()` variant. However the `connectSupplier` always [calls connect](https://github.com/reactor/reactor/blob/2.0.x/reactor-net/src/main/java/reactor/io/net/impl/netty/tcp/NettyTcpClient.java#L146) with the original address.

**Comment by rstoyanchev** [comment]
@shanexu and @romanek-adam there is a fix in 2.0.8 snapshots (commit https://github.com/reactor/reactor/commit/b7aa0e0f599f54425a5a3102d010849e9c041372) that should address this. It works on two levels. One, if you provide a host/port to begin with those will be re-resolved every time, but also the address returned from Reconnect should be now used.

It would be great to give this a try and confirm the fix. Thanks!

**Comment by romanek-adam** [comment]
A colleague of mine confirmed the fix. Therefore we're waiting for an official release of 2.0.8. Thanks!
